### DISCUSSION TITLE
Make a switch active only between certain times.

### DISCUSSION THREAD

**Discussion by homestead72** [arendst/Tasmota#14330] [Q&A] [answered]
I have a Shelly 1 flashed with the latest version of Tasmota connected to my bathroom fan.  The switch on the wall activates this via the SW input on the Shelly1.

However, the fan noise wakes up my son at night so I want to disable the Swtich1 between the hours of 10PM and 7AM even if the switch is pulled.

Is this possible?  I can't seem to find any rule examples that might help.

Thanks in advance

Stuart

**Comment by barbudor** [comment, marked as answer]
I would suggest :
- Using 2 timers to control the time at which switch is enable or disable (supposing timer 1 for disabling and 2 for enabing) => configured through web gui wiht action as Rule
- Matching rules 
```
   on clock#timer=1 do var1 0 endon
   on clock#timer=2 do var1 1 endon
```
- Then capture the switch changing to state ON (=1 - I'm assuming rocker switch with a switchmode of 1)
And send the command POWER with value %var1%
```
  on switch1#state=1 do power %var1% endon
```
So if switching ON during the disable timeframe, it will send a Power 0 which is not a problem since the fan should already be off
As the state=0 has not been captured, the switch back t OFF keeps is normal behavior do power off

**Reply by homestead72** [reply]
@barbudor  amazing.  It works like a dream.  Thank you and happy new Year.

**Reply by psantalucia** [reply]
@barbudor sorry for late question but maybe you can help me. Your rule and logic works great, i have esp32 with 4 relays and 4 switches controlling floor heating zones, switches attached to regular thermostats (just close a circuit till the desired temperature). I need at the end of disarm timeframe rule (on clock#timer=2 do…) to read switch states and set according relays to actual switch state, maybe during night (dissarm timeframe) some zones send signal or start heating and are in close switch state and if the switch is closed with these settings I need to open and close manually to reactive the relay. Thanks in advance

**Reply by barbudor** [reply]
You can't "read switch state". What you can do is to use a rule to capture the swich change into a variable.
The following should do what you are looking for:
```
   on clock#timer=1 do var1 0 endon
   on clock#timer=2 do backlog var1 1; power %var2% endon
   on switch1#state=0 do backlog var2 0; power 0 endon
   on switch1#state=1 do backlog var2 1; power %var1% endon
```
`var2` will always keep the last value of the switch.
When the timer2 occur to re-engage the normal behavior, it will set the power to the current position of the switch.

Note that the rule above will not switch off the power on timer1. If you want the power to go OFF when Timer1 is reached whatever is the state of the swicth, you could use:
```
   on clock#timer=1 do backlog var1 0; power 0 endon
   on clock#timer=2 do backlog var1 1; power %var2% endon
   on switch1#state=0 do backlog var2 0; power 0 endon
   on switch1#state=1 do backlog var2 1; power %var1% endon
```

Rules need to be adapted to your 4 relay/4 switches system by using the appropriate index on Switch and Power.
And using more variables as needed

**Reply by psantalucia** [reply]
@barbudor Thanks a lot!! It works perfect! I adapt the rules to 4 relay/4 switches and also add a start rule with:

`ON Switch1#Boot DO Backlog Power1 %value%; var2 %value%; var1 1 ENDON`

to initialize relays and variables, and everything is working great!

Thanks again for your help, i was stuck testing with the rules since thursday, also start reading about Berry scripting yesterday, but your help was really helpfull and appreciated. Regards

**Reply by psantalucia** [reply]
@barbudor last year you help me a lot with this rules to enable/dissable a relay in a time window and reset initial states after that, this is working great, today i found a small glitch that maybe you can help me, during the dissarm timeframe (some time after the rule: ON clock#timer=1 DO Backlog var1 0; Power1 0 ENDON) i suffer from a power outage in my house, so everything restarts whet power returns, the switch (analog thermostat) was in the ON state, so the relay was set to POWER ON and start heating the house, but it was during the "dissarm timeframe" and i need the relay to stay OFF. I think i need to improve the initialization of the variables and set var1 to 0 or 1 according to the correct timer timeframe, but don't know how to read or determine if the rule is in the arm timeframe or dissarm timeframe. Thanks a lot in advance

**Reply by sfromis** [reply]
What I prefer to do for heating is to boot with `PowerOnState 0` (always off), and then have logic to only power on if my logic detects that this is the right thing. That means that I do not really worry about timeframe for operations done before or around the restart. I'd rather be "a bit late" to power on, instead of power being on without the logic "expecting" that.

**Reply by psantalucia** [reply]
Thanks a lot!

**Comment by JacoAnnandale** [comment]
Good morning

I have done it as @barbudor  explained, but cant seem to get it working.

Can you please assist me ?

**Reply by JacoAnnandale** [reply]
AH got it  !!!

Changed my rule to  -  rule2 on **Power1#State=1**  do power %var1% endon .

But now, how do I evaluate conditions of timers on startup, to set the correct value of %var1%. Like initialize it ?

**Reply by barbudor** [reply]
What you are doing doesn't makes sense
My original rule was reacting to button/switch
You rule reacts to power changing from off to on
So there will be a glitch

Your use case may be different than the one from the original poster

You can't evaluate timers on startup
What you could is either:
- assume a default value in a 'on system#boot do..' rule
- or use rules based 'on time#minute' with < > comparators instead of timers. That's less convenient, doesn't easily supports sun rise/down options but works continuously

**Reply by JacoAnnandale** [reply]
Thanks for the reply

The reason my rule didnt work is because I used the button on GPIO0 (Push button) and not external switch , when I changed it to **Power1#State** the main Button (gpio0) responded as it should.

This is my scenario.  I want my electric blanket to activated ONLY between a certain time  ( even manually).  We get a lot of power cuts, and when power is restored, I want it to check if it is still in the allowed time slot and then set the power on state

In a nutshell I want to achieve the following 

1.  Sonoff can only be turned on during certain times    **Done, Thx**
2. When, booting, it must check if the ON and OFF time, and set the state accordingly
           - Because timers are ONESHot, it doesnt set the VAR1 on bootup

Hope  it makes better sense. My explanation previously was poor.

Thanks in advance

**Reply by barbudor** [reply]
If you want to trigger on Button instead of switch, the proper solution is to tribber on Button instead of Switch, not Power
It's a little more difficult as Button only provide a TOGGLE event, not a ON and OFF like a switch can do (depending on Switchmode)

Now to have something that works from boot up, you shouldn't use Timers (clock#timer=xxx) but a trigger based on minutes-passed-from-midnight `Time#Minute`

Assuming you want the button to be allowed only between 19:30 (== 1170 minutes from midnight) and 22:30 (== 1350 min from midnight), the following rules should provide what you need:
```
Rule1 
  on time#initialized do event allow %value%
  on time#minute do event allow %value%
  on event#allow<1170 do backlog var1 0; power off break
  on event#allow<1350 do var1 2 break
  on event#allow>=1350 do backlog var1 0; power off break
  on button1#state do power %var1% endon
```
The 1st trigger will be executed as soon as NTP has synced the time and provide the number of minutes from midnight to the event "allow"
The 2nd trigger will provide the number of minutes from midnight to the event "allow"
The 3 following triggers on the event "allow" will decide the value of var1 between 0 (== off) and 2 (== toggle), enforcing a power OFF for the 2 cases outside the allowed period
The last trigger reacts on the button, allowing the button to perform it function (toggle the relay) only within the allowed time period

That doesn't make use of the timers so it is a bit less convenient to configure and update but, it's the only solution to have something that works straight at any reboot

Note that will generate a lot of logs in the console (and MQTT message) every minute
If you want to avoid those message, you can combine Timers and Time#Minute but that will be even more complicated to update as you will need to update both Timers and Rules in case you want to change your trigger time:
```
Rule1 
  on time#initialized<1170 do backlog var1 0; power off break
  on time#initialized<1350 do var1 2 break
  on time#initialized>=1350 do backlog var1 0; power off break
  on clock#timer=1 do var1 2 endon
  on clock#timer=2 do backlog var1 0; power off endon
  on button1#state do power %var1% endon
```

**Reply by JacoAnnandale** [reply]
Wow, that is awesome!!   Thanks a lot.  Will give it a go and report back.

**Reply by Littlesloths** [reply]
This post helped me a lot. Thank you!

Still there seems to be a fault: it takes: allow=%value%
without "=" the allow<xyz will never trigger
Also, I am wondering if two endon are missing:

Rule1 
  on time#initialized do event allow=%value% endon
  on time#minute do event allow=%value% endon
  on event#allow<1170 do backlog var1 0; power off break
  on event#allow<1350 do var1 2 break
  on event#allow>=1350 do backlog var1 0; power off break
  on button1#state do power %var1% endon

**Reply by sfromis** [reply]
You are right that the `Event` command needs `=` between event name and value, and every rule must indeed be `on...do...endon` or `on...do...break`

As I see the post from barbudor, it was mainly to illustrate the a feasible technique, without being proofread for syntax details.

**Reply by Littlesloths** [reply]
Hey, first, as I said, the post very useful for me. I appreciate the effort a lot, especially what you @sfromis and @barbudor are doing!

Still, it took me several hours to find the root cause why it was not working. Therefore, I just wanted to give the result back to the community to be useful somehow. I guess you know that the code samples from the topic here and elsewhere is helping thousands of others... Note that several KI LLMs use github as a training ressource too. I was impressed how good e.g. ChatGPT could help me finding the issue, but at this specific matter ChatGPT was not correct. Now I hope there is some input and feedback to confirm the result working.
